src/workflow/invocation.rs

```diff
@@ -31,85 +34,82 @@ impl Into<Ipld> for Invocation {
 }
 
 impl TryFrom<Ipld> for Invocation {
-    type Error = ();
+    type Error = anyhow::Error;
 
     fn try_from(ipld: Ipld) -> Result<Self, Self::Error> {
         match ipld {
-            Ipld::Map(assoc) => {
-                let run: Batch = match assoc.get("run") {
-                    Some(ipld) => Batch::try_from(ipld.clone()),
-                    _ => Err(()),
-                }?;
-
-                let meta = match assoc.get("meta") {
-                    Some(ipld) => ipld.clone(),
-                    None => Ipld::Null,
-                };
-
-                let prf = match assoc.get("prf") {
+            Ipld::Map(assoc) => Ok(Invocation {
+                meta: assoc.get("meta").map(Clone::clone).unwrap_or(Ipld::Null),
+                run: assoc
+                    .get("run")
+                    .ok_or(anyhow!("run field is empty"))
+                    .and_then(Batch::try_from)
+                    .unwrap(),
+                prf: match assoc.get("prf") {
                     Some(Ipld::List(vec)) => {
                         vec.iter().try_fold(Vec::new(), |mut acc, ipld| match ipld {
                             Ipld::Link(cid) => {
                                 acc.push(Link::new(*cid));
                                 Ok(acc)
                             }
-                            _ => Err(()),
+                            _ => bail!("Not a link"),
                         })
                     }
-                    _ => Err(()),
-                }?;
-
-                Ok(Invocation { meta, prf, run })
-            }
-            _ => Err(()),
+                    other => bail!("Expected a List, but got something else: {:?}", other),
```

Comment:
I threw a bunch of `{:?}` in because the conversions to normal strings was... not as simple as I wanted. This is probably overkill 🤷‍♀️